lib/rex/ui/text/dispatcher_shell.rb

```diff
@@ -359,11 +359,11 @@ def initialize(prompt, prompt_char = '>', histfile = nil, framework = nil)
   #
   def tab_complete(str)
     # Check trailing whitespace so we can tell 'x' from 'x '
-    str_match = str.match(/\s+$/)
+    str_match = str.match(/[^\\]([\\]{2})*\s+$/)
     str_trail = (str_match.nil?) ? '' : str_match[0]
 
     # Split the line up by whitespace into words
-    str_words = str.split(/[\s\t\n]+/)
+    str_words = Shellwords.split(str)
```

Comment:
I'm glad to see more `Shellwords` in our bespoke console, but this is more sensitive to input than `split` and will trigger an `ArgumentError` on unbalanced quotes, for instance.

```
msf5 exploit(unix/webapp/drupal_drupalgeddon2) > options

Module options (exploit/unix/webapp/drupal_drupalgeddon2):

   Name         Current Setting  Required  Description
   ----         ---------------  --------  -----------
   DUMP_OUTPUT  false            no        Dump payload command output
   PHP_FUNC     passthru         yes       PHP function to execute
   Proxies                       no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS                        yes       The target address range or CIDR identifier
   RPORT        80               yes       The target port (TCP)
   SSL          false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI    /                yes       Path to Drupal install
   VHOST                         no        HTTP server virtual host


Exploit target:

   Id  Name
   --  ----
   0   Automatic (PHP In-Memory)


msf5 exploit(unix/webapp/drupal_drupalgeddon2) > show targets

Exploit targets:

   Id  Name
   --  ----
   0   Automatic (PHP In-Memory)
   1   Automatic (PHP Dropper)
   2   Automatic (Unix In-Memory)
   3   Automatic (Linux Dropper)
   4   Drupal 7.x (PHP In-Memory)
   5   Drupal 7.x (PHP Dropper)
   6   Drupal 7.x (Unix In-Memory)
   7   Drupal 7.x (Linux Dropper)
   8   Drupal 8.x (PHP In-Memory)
   9   Drupal 8.x (PHP Dropper)
   10  Drupal 8.x (Unix In-Memory)
   11  Drupal 8.x (Linux Dropper)


msf5 exploit(unix/webapp/drupal_drupalgeddon2) > set target Automatic\ (PHP\ Dropper)
target => Automatic (PHP Dropper)
msf5 exploit(unix/webapp/drupal_drupalgeddon2) > options

Module options (exploit/unix/webapp/drupal_drupalgeddon2):

   Name         Current Setting  Required  Description
   ----         ---------------  --------  -----------
   DUMP_OUTPUT  false            no        Dump payload command output
   PHP_FUNC     passthru         yes       PHP function to execute
   Proxies                       no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS                        yes       The target address range or CIDR identifier
   RPORT        80               yes       The target port (TCP)
   SSL          false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI    /                yes       Path to Drupal install
   VHOST                         no        HTTP server virtual host


Exploit target:

   Id  Name
   --  ----
   1   Automatic (PHP Dropper)


msf5 exploit(unix/webapp/drupal_drupalgeddon2) > set target "AuTraceback (most recent call last):
	20: from ./msfconsole:49:in `<main>'
	19: from /rapid7/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start'
	18: from /rapid7/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'
	17: from /rapid7/metasploit-framework/lib/rex/ui/text/shell.rb:134:in `run'
	16: from /rapid7/metasploit-framework/lib/rex/ui/text/shell.rb:309:in `get_input_line'
	15: from /rapid7/metasploit-framework/lib/rex/ui/text/input/readline.rb:100:in `pgets'
	14: from /rapid7/metasploit-framework/lib/rex/ui/text/input/readline.rb:162:in `readline_with_output'
	13: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:4875:in `readline'
	12: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:4853:in `readline_internal'
	11: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:4779:in `readline_internal_charloop'
	10: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:4363:in `_rl_dispatch'
	 9: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:4374:in `_rl_dispatch_subseq'
	 8: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:6903:in `rl_complete'
	 7: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:6813:in `rl_complete_internal'
	 6: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/rbreadline.rb:6329:in `gen_completion_matches'
	 5: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rb-readline-0.5.5/lib/readline.rb:136:in `readline_attempted_completion_function'
	 4: from /rapid7/metasploit-framework/lib/rex/ui/text/shell.rb:62:in `block in init_tab_complete'
	 3: from /rapid7/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:366:in `tab_complete'
	 2: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/2.6.0/shellwords.rb:82:in `shellsplit'
	 1: from /Users/wvu/.rbenv/versions/2.6.2/lib/ruby/2.6.0/shellwords.rb:82:in `scan'
/Users/wvu/.rbenv/versions/2.6.2/lib/ruby/2.6.0/shellwords.rb:84:in `block in shellsplit': Unmatched double quote: "set target \"Au" (ArgumentError)
wvu@kharak:/rapid7/metasploit-framework:HEAD$
```

Reply:
Whoops yeah we definitely can't have any of that. I'll figure out how to fix this and push a new version. Great catch @wvu-r7 :+1: 

Are there other instances you can think of where an `ArgumentError` would be raised? I'll look through the `Shellwords` docs of course, but if you know of something else I'll be sure to handle it.

Reply:
I think it's only unmatched quotes. It's not just double quotes, so the error message is a lie.

https://github.com/ruby/ruby/blob/trunk/lib/shellwords.rb
https://github.com/ruby/ruby/blob/trunk/test/test_shellwords.rb

Reply:
So I implemented a custom `shellsplit` function, which when unbalanced quotes are present, will treat the rest of the string as if the last character were the closing quote. Both the words and the unbalanced quote (if present) are then returned in a hash. The quote then get's passed through to `tab_complete_stub` where it's prepended to words and the words do not have their spaces escaped.

Should work quite nicely now for your `set TARGET "Au` example.

Reply:
Wow!!